### PR TITLE
13203 Adds documentation for Account API

### PR DESCRIPTION
**Pull request by Hronom**
Extend OpenAPI documentation with Account API.

This should help with this https://github.com/keycloak/keycloak/issues/13203 but there some nuances.

# How-to check
Generate schema:
```
mvn clean install -DskipTests=true smallrye-open-api:generate-schema -P jboss-release -pl services
```

Check generated file in:
```
${project.build.directory}/apidocs-rest/swagger/apidocs/openapi.yaml
```

# Nuances

There [issue related to how API organized](https://github.com/keycloak/keycloak/issues/13203#issuecomment-1876220056).
Specifically there sub resources here:
```
    @Path("/")
    public Object getAccountService() {
```
and here:
```
    @Path("{version : v\\d[0-9a-zA-Z_\\-]*}")
    @Produces(MediaType.APPLICATION_JSON)
    public Object getVersionedAccountRestService(final @PathParam("version") String version) {
```

During generation of OpenAPI plugin `smallrye-open-api-maven-plugin` not able to to get a clue about what Sub Resource will be returned by was methods.

So in order to provide at least something instead of nothing(for Account API) - there was added groups of fake methods in `AccountLoader`.
```
    /**
     * This needed to provide examples for Account API.
     * Here we return a specific sub resource, so OpenAPI generator able to scan it.
     */
    @Path("/account-api-cors-example")
    public CorsPreflightService getAccountApiCorsExample() {
        throw new ForbiddenException();
    }

    /**
     * This needed to provide examples for Account API.
     * Here we return a specific sub resource, so OpenAPI generator able to scan it.
     */
    @Path("/account-api-example")
    public AccountRestService getAccountApiExample() {
        throw new ForbiddenException();
    }
```

```
    /**
     * This needed to provide examples for Account API.
     * Here we return a specific sub resource, so OpenAPI generator able to scan it.
     */
    @Path("{version : v\\d[0-9a-zA-Z_\\-]*}/account-api-cors-example")
    @Produces(MediaType.APPLICATION_JSON)
    public CorsPreflightService getVersionedAccountApiCorsExample(final @PathParam("version") String version) {
        throw new ForbiddenException();
    }

    /**
     * This needed to provide examples for Account API.
     * Here we return a specific sub resource, so OpenAPI generator able to scan it.
     */
    @Path("{version : v\\d[0-9a-zA-Z_\\-]*}/account-api-example")
    @Produces(MediaType.APPLICATION_JSON)
    public AccountRestService getVersionedAccountApiExample(final @PathParam("version") String version) {
        throw new ForbiddenException();
    }
```

This allow `smallrye-open-api-maven-plugin` to generate OpenAPI spec. However it generates them under `*-example` paths.

To make it actual we need to do simple replace of `/account-api-cors-example` and `/account-api-example` to empty string. This helps developers to easily adopt generated OpenAPI for other services that need to use Account API.

### CODE REVIEW

**Comment by Hronom** [comment]
cc @ahus1 

I'm totally ok if maintainers edit this PR.

Merging this PR helps a lot to others who need to work with Account API e.g. create custom Account Console

File: services/src/main/java/org/keycloak/services/resources/account/AccountLoader.java [hunk 1]
```diff
@@ -16,8 +16,11 @@
  */
 package org.keycloak.services.resources.account;
 
+import jakarta.ws.rs.*;
```
Comment (thomasdarimont) [inline services/src/main/java/org/keycloak/services/resources/account/AccountLoader.java:19]: Nitpick: avoid *-imports

File: services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java [hunk 2]
```diff
@@ -30,21 +30,18 @@
...
-import jakarta.ws.rs.DefaultValue;
-import jakarta.ws.rs.GET;
-import jakarta.ws.rs.NotFoundException;
-import jakarta.ws.rs.POST;
-import jakarta.ws.rs.PUT;
-import jakarta.ws.rs.Path;
-import jakarta.ws.rs.PathParam;
-import jakarta.ws.rs.Produces;
-import jakarta.ws.rs.QueryParam;
+import jakarta.ws.rs.*;
```
Comment (thomasdarimont) [inline services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java:33]: Nitpick: Avoid *-imports

File: services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java [hunk 3]
```diff
@@ -440,6 +480,8 @@ public Stream<GroupRepresentation> groupMemberships(@QueryParam("briefRepresenta
     @GET
     @Produces(MediaType.APPLICATION_JSON)
     @NoCache
+    @Tag(name = KeycloakOpenAPI.Admin.Tags.ACCOUNT)
+    @Operation(summary = "Applications.")
```
Comment (thomasdarimont) [inline services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java:484]:
```suggestion
    @Operation(summary = "List applications available to the user.")
```

File: services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java [hunk 4]
```diff
@@ -430,6 +468,8 @@ public LinkedAccountsResource linkedAccounts() {
     @GET
     @NoCache
     @Produces(MediaType.APPLICATION_JSON)
+    @Tag(name = KeycloakOpenAPI.Admin.Tags.ACCOUNT)
+    @Operation(summary = "Group memberships.")
```
Comment (thomasdarimont) [inline services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java:472]:
```suggestion
    @Operation(summary = "List group memberships.")
```

File: services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java [hunk 5]
```diff
@@ -149,6 +148,8 @@ public UserRepresentation account(final @QueryParam("userProfileMetadata") Boole
     @Consumes(MediaType.APPLICATION_JSON)
     @Produces(MediaType.APPLICATION_JSON)
     @NoCache
+    @Tag(name = KeycloakOpenAPI.Admin.Tags.ACCOUNT)
+    @Operation(summary = "Update account.")
```
Comment (thomasdarimont) [inline services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java:152]:
```suggestion
    @Operation(summary = "Update account information.")
```

**Review by thomasdarimont — CHANGES_REQUESTED** [review]
Thanks for your PR! Changes look good to me, just a few nitpicks :)

**Review by stianst — CHANGES_REQUESTED** [review]
This would be great to get in, but it needs a bit of work.

We should not have a single docs for both account and admin I'd say, but rather have two separate docs.

The annotations here also need some more review and tuning.

For example this is not the right URL (it should be /realms/{realm}/account)
image

Another example is CORS preflight stuff shouldn't be listed I'd say:
image

If you're still wanting to work on this, I'd be happy to review it in more depth and can also work on the website side of things if you want.

**Comment by Hronom** [comment]
@thomasdarimont I'm ok if maintainers will do all changes that they think we need to do.
I wanna make sure that this PR will be not forgotten.

@stianst I wrote reasons in description about why it's done like this. I consider this ok solutions, since right now keycloak don't have any documentation for `/accounts` API's and it requested by many people. So instead of nothing I select something.

Maintainers, please adjust whatever you think to need to adjust. Let's not put this changes in forgotten mode.
I'm providing my verbal OK to do any changes in this PR, or feel free to create new one.

**Comment by Hronom** [comment]
@stianst you probably really out of context and didn't read description of the PR. 
The way how Keycloack organize API there no possibility to make paths:
```
/accounts
```

And I would say this is the reasons why Accounts API still don't have documentation, because whenever maintainers of Keycloack tried to do it the hit same wall as I and drop work on it.

There was done CRITICAL architecture mistake in the beginning when Accounts API was created. 
Probably persons who created that was not care too much about it.
Unfortunately the documentation libraries can't support generic API's.

**Pull request closed without merging**